src/mono/wasm/runtime/hybrid-globalization/grapheme-segmenter.ts

```diff
@@ -0,0 +1,138 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+/**
+ * This file is partially using code from FormatJS Intl.Segmenter implementation, reference: 
+ * https://github.com/formatjs/formatjs/blob/58d6a7b398d776ca3d2726d72ae1573b65cc3bef/packages/intl-segmenter/src/segmenter.ts
+ * https://github.com/formatjs/formatjs/blob/58d6a7b398d776ca3d2726d72ae1573b65cc3bef/packages/intl-segmenter/src/segmentation-utils.ts
+ */
+
+import { SegmentationRules } from "./segmentation-rules";
+
+type SegmentationRule = {
+    breaks: boolean
+    before?: RegExp
+    after?: RegExp
+}
+
+type SegmentationRuleRaw = {
+    breaks: boolean
+    before?: string
+    after?: string
+}
+  
+type SegmentationTypeRaw = {
+    variables: Record<string, string>
+    rules: Record<string, SegmentationRuleRaw>
+}
+
+function is_surrogate_pair(str: string, index: number): boolean {
+    const high = str.charCodeAt(index - 1);
+    const low = str.charCodeAt(index);
+
+    return 0xD800 <= high && high <= 0xDBFF && 0xDC00 <= low && low <= 0xDFFF;
+}
+
+function replace_variables(variables: Record<string, string>, input: string): string {
+    const findVarRegex = /\$[A-Za-z0-9_]+/gm;
+    return input.replaceAll(findVarRegex, match => {
+        if (!(match in variables)) {
+            throw new Error(`No such variable ${match}`);
+        }
+        return variables[match];
+    });
+}
+
+function generate_rule_regex (rule: string, variables: Record<string, string>, after: boolean): RegExp {
+    return new RegExp(`${after ? "^" : ""}${replace_variables(variables, rule)}${after ? "" : "$"}`);
+}
+
+export class GraphemeSegmenter {
+    private readonly rules;
+    private readonly ruleSortedKeys;
+
+    public constructor() {  
+        // Process segmentation rules
+        this.rules = GraphemeSegmenter.prepare_segmanation_rules(SegmentationRules);
```

Comment:
It looks like every time we call index_of etc, we create a new GraphemeSegmenter instance, which means this initialization logic runs for every operation. Can we avoid that, by reusing a segmenter or something?

Reply:
Yes, you are right. I added global variable `graphemeSegmenterCached` and only process segmentation rules on first call https://github.com/dotnet/runtime/blob/0ea511854ba5ca19bec62e711264cacfb45ef60d/src/mono/wasm/runtime/hybrid-globalization/collations.ts#L121. Let me know if this is what you had in mind.

Reply:
Perfect.